components/eamxx/src/physics/p3/impl/p3_init_impl.hpp

```diff
@@ -3,35 +3,327 @@
+  // proportional to qr/Nr and shape parameter mu_r
+  for (ii = 1; ii <= 10; ++ii) {
+    mu_r = 1; // mu_r_constant = 1
+
+    // loop over number-weighted mean size
+    for (jj = 1; jj <= 300; ++jj) {
+      if (jj <= 20) {
+        dm = (jj*10 - 5)*1.e-6; // mean size [m]
+      }
+      else {
+        dm = ((jj-20)*30 + 195)*1.e-6; // mean size [m]
+      }
+
+      lamr = (mu_r + 1)/dm;
+
+      // do numerical integration over PSD
+
+      dum1 = 0; // numerator,   number-weighted fallspeed
+      dum2 = 0; // denominator, number-weighted fallspeed
+      dum3 = 0; // numerator,   mass-weighted fallspeed
+      dum4 = 0; // denominator, mass-weighted fallspeed
+      dum5 = 0; // term for ventilation factor in evap
+      dd   = 2;
+
+      // loop over PSD to numerically integrate number and mass-weighted mean fallspeeds
+      for (kk = 1; kk <= 10000; ++kk) {
+
+        dia = (kk*dd - dd*0.5)*1.e-6;     // size bin [m]
+        amg = c::PIOV6*997 * std::pow(dia, 3); // mass [kg]
+        amg = amg*1000;                   // convert [kg] to [g]
+
+        // get fallspeed as a function of size [m s-1]
+        if (dia*1.e+6 <= 134.43) {
+          vt = 4.5795e+3 * std::pow(amg, 2*thrd);
+        }
+        else if (dia*1.e+6 < 1511.64) {
+          vt = 4.962e+1 * std::pow(amg, thrd);
+        }
+        else if (dia*1.e+6 < 3477.84) {
+          vt = 1.732e+1 * std::pow(amg, c::SXTH);
+        }
+        else {
+          vt = 9.17;
+        }
+
+        // note: factor of 4.*mu_r is non-answer changing and only needed to
+        //       prevent underflow/overflow errors, same with 3.*mu_r for dum5
+        dum1 += vt * std::pow(10, mu_r*std::log10(dia) + 4*mu_r) * std::exp(-lamr*dia) * dd * 1.e-6;
+        dum2 += std::pow(10, mu_r*std::log10(dia) + 4*mu_r) * std::exp(-lamr*dia) * dd * 1.e-6;
+        dum3 += vt * std::pow(10, (mu_r+3)*std::log10(dia) + 4*mu_r) * std::exp(-lamr*dia) * dd * 1.e-6;
+        dum4 += std::pow(10, (mu_r+3)*std::log10(dia) + 4*mu_r) * std::exp(-lamr*dia) * dd * 1.e-6;
+        dum5 += std::pow(vt*dia, 0.5) * std::pow(10, (mu_r+1)*std::log10(dia) + 3*mu_r) * std::exp(-lamr*dia) * dd * 1.e-6;
+      }
+
+      dum2 = std::max(dum2, 1.e-30); // to prevent divide-by-zero below
+      dum4 = std::max(dum4, 1.e-30); // to prevent divide-by-zero below
+      dum5 = std::max(dum5, 1.e-30); // to prevent log10-of-zero below
+
+      vn_table_vals_nc(jj-1,ii-1)    = dum1/dum2;
+      vm_table_vals_nc(jj-1,ii-1)    = dum3/dum4;
+      revap_table_vals_nc(jj-1,ii-1) = std::pow(10, std::log10(dum5) + (mu_r+1)*std::log10(lamr) - (3*mu_r));
+    }
+  }
+
+  mu_r_table_vals = mu_r_table_vals_nc;
+  vn_table_vals = vn_table_vals_nc;
+  vm_table_vals = vm_table_vals_nc;
+  revap_table_vals = revap_table_vals_nc;
+}
+
+template <bool IsRead>
+struct IoAction
+{
+  template <typename S>
+  static void action(const ekat::FILEPtr& fid, S* data, const size_t size)
```

Comment:
Up to you, but since we are doing C++ only, maybe we should just use `<fstream>` stuff, rather than the cumbersome `FILEPtr`? It seems easier to understand for new devs, and easier to maintain.